qiskit/primitives/containers/bit_array.py

```diff
@@ -347,3 +372,216 @@ def reshape(self, *shape: ShapeInput) -> "BitArray":
+        """Join a sequence of bit arrays along shots.
+
+        Args:
+            bitarrays: The bit arrays must have (1) the same number of bits,
+                and (2) the same shape.
+
+        Returns:
+            BitArray: The stacked bit array.
+
+        Raises:
+            ValueError: If the sequence of bit arrays is empty.
+            ValueError: If any bit arrays has a different number of bits.
+            ValueError: If any bit arrays has a different shape.
+        """
+        if len(bitarrays) == 0:
+            raise ValueError("Need at least one bit array to stack")
+        num_bits = bitarrays[0].num_bits
+        shape = bitarrays[0].shape
+        for i, ba in enumerate(bitarrays):
+            if ba.num_bits != num_bits:
+                raise ValueError(
+                    "All bit arrays must have same number of bits, "
+                    f"but the bit array at index 0 has {num_bits} bits "
+                    f"and the bit array at index {i} has {ba.num_bits} bits."
+                )
+            if ba.shape != shape:
+                raise ValueError(
+                    "All bit arrays must have same shape, "
+                    f"but the bit array at index 0 has shape {shape} "
+                    f"and the bit array at index {i} has shape {ba.shape}."
+                )
+        data = np.concatenate([ba.array for ba in bitarrays], axis=-2)
+        return BitArray(data, num_bits)
+
+    @staticmethod
+    def stack_bits(bitarrays: Sequence[BitArray]) -> BitArray:
+        """Join a sequence of bit arrays along bits.
+
+        Args:
+            bitarrays: The bit arrays must have (1) the same number of shots,
+                and (2) the same shape.
+
+        Returns:
+            BitArray: The stacked bit array.
+
+        Raises:
+            ValueError: If the sequence of bit arrays is empty.
+            ValueError: If any bit arrays has a different number of shots.
+            ValueError: If any bit arrays has a different shape.
+        """
+        if len(bitarrays) == 0:
+            raise ValueError("Need at least one bit array to stack")
+        num_shots = bitarrays[0].num_shots
+        shape = bitarrays[0].shape
+        for i, ba in enumerate(bitarrays):
+            if ba.num_shots != num_shots:
+                raise ValueError(
+                    "All bit arrays must have same number of shots, "
+                    f"but the bit array at index 0 has {num_shots} shots "
+                    f"and the bit array at index {i} has {ba.num_shots} shots."
+                )
+            if ba.shape != shape:
+                raise ValueError(
+                    "All bit arrays must have same shape, "
+                    f"but the bit array at index 0 has shape {shape} "
+                    f"and the bit array at index {i} has shape {ba.shape}."
+                )
+        data = np.concatenate([_unpack(ba) for ba in bitarrays], axis=-1)
+        data, num_bits = _pack(data)
```

Comment:
Obvious to you I'm sure, but we can, in the future, work to eliminate the temporary 8x memory overhead here.

Reply:
Yes, we need to think of the way to reduce the memory overhead. If it is hard to do it with numpy, rust would be an option.